tests/test_callbackcontext.py

```diff
@@ -58,6 +58,20 @@ def test_from_job(self, app):
         assert callback_context.job_queue is app.job_queue
         assert callback_context.update_queue is app.update_queue
 
+    def test_job_queue(self, bot, app, recwarn):
+        expected_warning = (
+            "No `JobQueue` set up. To use `JobQueue`, you must install PTB via "
+            "`pip install python-telegram-bot[job_queue]`."
+        )
+
+        callback_context = CallbackContext(app)
+        assert callback_context.job_queue is app.job_queue
+        app = ApplicationBuilder().job_queue(None).token(bot.token).build()
+        callback_context = CallbackContext(app)
+        assert callback_context.job_queue is None
+        assert len(recwarn) == 1
+        assert str(recwarn[0].message) == expected_warning
```

Comment:
looks good overall :)

```suggestion
        assert str(recwarn[0].message) == expected_warning
        assert recwarn[0].filename == __file__, "wrong stacklevel"
```

The additional assertion makes sure that the warning is issued directly from the line where the user uses `context.job_queue` so that they are pointed to the right place.

Please also add an analogous test directly to `TestApplication` where the warning behavior of `Application.job_queue` should be tested as well

Reply:
IISC this is failing the tests :thinking:

Reply:
Then the stacklevel is indeed wrong :) The `stacklevel` parameter of `warn` does the following. When an update is handled, roughly the following methods are called:

`App.process_update`, which calls `BaseHandler.handle_update`, which calls the users `callback` in which the user accesses `context.job_queue`, which accesses the property `Application.job_queue`, which calls `_utils.warn` which issues the warning.

Now by default the warning is issued directly from `tg._utils.warn` and that filename will be shown to the user as the origin of the warning. But this is not helpful for the user at all b/c the user didn't manually call `_utils.warn`. Passing a higher value to `stacklevel` tells python to go up in the "call stack" and we want to pass a value such that the accessing of `context.job_queue` is shown as origin of the warning :)

Reply:
Wow, thanks for the explanation!